src/EditorFeatures/Core/Implementation/GoToImplementation/GoToImplementationCommandHandler.cs

```diff
@@ -0,0 +1,113 @@
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.GoToImplementation
+{
+    [ExportCommandHandler(PredefinedCommandHandlerNames.GoToImplementation,
+        ContentTypeNames.RoslynContentType)]
+    internal sealed class GoToImplementationCommandHandler : ICommandHandler<GoToImplementationCommandArgs>
+    {
+        private readonly IWaitIndicator _waitIndicator;
+        private readonly IEnumerable<Lazy<INavigableItemsPresenter>> _navigatableItemPresenters;
+
+        [ImportingConstructor]
+        public GoToImplementationCommandHandler(
+            [ImportMany] IEnumerable<Lazy<INavigableItemsPresenter>> navigatableItemPresenters,
+            IWaitIndicator waitIndicator)
+        {
+            _navigatableItemPresenters = navigatableItemPresenters;
+            _waitIndicator = waitIndicator;
+        }
+
+        public CommandState GetCommandState(GoToImplementationCommandArgs args, Func<CommandState> nextHandler)
+        {
+            // Because this is expensive to compute, we just always say yes
+            return CommandState.Available;
+        }
+
+        public void ExecuteCommand(GoToImplementationCommandArgs args, Action nextHandler)
+        {
+            var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+
+            if (caret.HasValue)
+            {
+                var document = args.SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document != null)
+                {
+                    // We have all the cheap stuff, so let's do expensive stuff now
+                    string messageToShow = null;
+                    _waitIndicator.Wait(
+                        EditorFeaturesResources.GoToImplementationTitle,
+                        EditorFeaturesResources.GoToImplementationMessage,
+                        allowCancel: true,
+                        action: context => messageToShow = ExecuteCommandCoreAndReturnMessage(document, caret.Value, context.CancellationToken));
+
+                    if (messageToShow != null)
+                    {
+                        var notificationService = document.Project.Solution.Workspace.Services.GetService<INotificationService>();
+                        notificationService.SendNotification(messageToShow,
+                            title: EditorFeaturesResources.GoToImplementationTitle,
+                            severity: NotificationSeverity.Information);
+                    }
+
+                    return;
+                }
+            }
+
+            nextHandler();
+        }
+
+        private string ExecuteCommandCoreAndReturnMessage(Document document, int position, CancellationToken cancellationToken)
+        {
+            var symbol = SymbolFinder.FindSymbolAtPositionAsync(document, position, cancellationToken).WaitAndGetResult(cancellationToken);
```

Comment:
This won't work for TypeScript.  Could you instead MEF import an appropriate language service for this?